src/dbus/mod.rs

```diff
@@ -0,0 +1,94 @@
+//! D-Bus server interface actor.
+
+use crate::update_agent::{LastRefresh, UpdateAgent};
+use actix::prelude::*;
+use actix::Addr;
+use core::convert::TryFrom;
+use futures::executor;
+use futures::prelude::*;
+use log::trace;
+use zbus::{dbus_interface, fdo};
+use zvariant::ObjectPath;
+
+/// Experimental interface for testing.
+struct Experimental {
+    agent_addr: Addr<UpdateAgent>,
+}
+
+#[dbus_interface(name = "org.coreos.zincati1.Experimental")]
+impl Experimental {
+    /// Just a test method.
+    fn moo(&self, talkative: bool) -> String {
+        if talkative {
+            String::from("Moooo mooo moooo!")
+        } else {
+            String::from("moo.")
+        }
+    }
+
+    /// Get update_agent actor's last refresh time.
+    fn last_refresh_time(&self) -> Result<i64, zbus::fdo::Error> {
+        let msg = LastRefresh {};
+        let refresh_time_fut = self
+            .agent_addr
+            .send(msg)
+            .unwrap_or_else(|e| Err(e.into()))
+            .map_err(|e| {
+                let err_msg = format!("failed to get last refresh time from agent actor: {}", e);
+                log::error!("{}", err_msg);
+                zbus::fdo::Error::Failed(err_msg)
```

Comment:
I think it's generally better to "log xor propagate" i.e. don't both log and propagate an error, because that usually leads to double-logging at least (and if the *callers* also log and propagate you can get many copies).

Although this case is tricky because we're also sending an error to the client, but it's useful to log in the journal too.  So...this is probably OK; it's what rpm-ostree does too.

Reply:
Thanks for pointing that out! But just for my information, in this case, do we still count it as "propagating"? My intuition is that here we are sending a message to a client that an error occurred; but on the server side, the error was never really propagated anywhere, and wouldn't be double logged elsewhere right?

Reply:
It makes sense here - the error is being returned to the client, but we don't control their logging, and in some cases the admin would like to be able to reliably see errors from the journal.

So fine as is!